### DISCUSSION TITLE
Global Variables from File Output Node and Video Editing Render

### DISCUSSION THREAD

**Discussion by Fiztban** [p2or/blender-loom#65] [Ideas]
Hello,

I have stumbled upon your addon and found it really useful as its the first implementation I have found of variables in output file names that has enough versatility!

However I would like to propose some enhancements given my use case where the benefits of your addon cannot be used to their fullest.

I have several shots that I wish to render of objects with multiple camera angles. These are for creating standardised object renders (a ball and a cube in my example). To make this work each object is set up in its own scene:

![image](https://user-images.githubusercontent.com/30956188/96242185-11540a80-0f9b-11eb-8409-c7aa8052ac6d.png)

I set up each object in a separate scene because it allows me to bind cameras of the individual views to markers (Ctrl+B when camera is selected) so that I can set up a render sequence for all the required views of each object (within a scene):

![image](https://user-images.githubusercontent.com/30956188/96242654-9ccd9b80-0f9b-11eb-9072-fe00e078cf76.png)

Within each scene I am able to specify the output using loom which is fantastic, because the scene can hold the name of the object, and the cameras the names of the views:

![image](https://user-images.githubusercontent.com/30956188/96243041-18c7e380-0f9c-11eb-980a-1f0ffcb24a9d.png)

Up to here it is everything I hoped for. The cameras bound to markers actually register correctly in the Global variables when saving the frame names:

![image](https://user-images.githubusercontent.com/30956188/96244575-2b431c80-0f9e-11eb-8a4e-a01bcb4c17b4.png)

However, part of the workflow I am trying to establish is to have the ability to re-render each shot for either 1 object, or ALL objects. To achieve this I simply create a new "All Renders" scene and move to the Video Editing Sequencer. There I add in all the other Object Scenes one after the other so that all views can be rendered one after the next:

![image](https://user-images.githubusercontent.com/30956188/96243283-6e9c8b80-0f9c-11eb-9460-90f29b582399.png)

![image](https://user-images.githubusercontent.com/30956188/96243304-79efb700-0f9c-11eb-8b17-33df4632513f.png)

This is the point at which loom's functionality breaks down. The moment I use the Render Image Sequence (CTRL+SHIFT+F12) loom proceeds to render the correct number of frames...

![image](https://user-images.githubusercontent.com/30956188/96243630-e79be300-0f9c-11eb-9383-f8b48fb09c41.png)

..but then simply reports the error that no cameras can be found within the scene:

![image](https://user-images.githubusercontent.com/30956188/96243683-f97d8600-0f9c-11eb-8a3e-5dd08d2a417d.png)

It would be amazing if the already solid Global variables functionality of loom could be extended to work in this case.

Lastly if the Global variables functionality can also be extended not only to work within the video sequencer when the scenes are being called from an external scene, but also to work with the File Output node file path system, I think this would largely solve many people's desires for variables within filenames.

The reason I would personally want Global variables within the file output nodes is because I would actually prefer to save each object's render in more than one place as various people would end up using them.

I hope this is a suggestion that is implementable as I believe many people would find it incredibly useful.

**Comment by p2or** [comment]
Hi @Fiztban,

thanks for your suggestion. I'm working on it and let you know whether that's going to work.

Cheers,
Christian

**Comment by Fiztban** [comment]
@p2or ,

Thank you for letting me know. If there are any other details you might require I'm happy to do my best to assist.

Good luck and I look forward to any success you may find.

**Comment by RyIock** [comment]
![image](https://user-images.githubusercontent.com/115088138/206216594-1ecb947d-11e3-439c-ad06-bbb8ce48a031.png)

Is it possible to remove these numbers? I'm using the $MARKER to give these renders different naming scemes but I dont need the 0001 at the end or any number.

**Comment by p2or** [comment]
Hi @RyIock,

this is currently not possible, can be implemented though. I'd recommend open a new issue and explain in detail why this feature would be nice to have, keeps it organized here as well.

Cheers,
Christian

**Comment by nenad-jalsovec** [comment]
Hello,

Great addon! 
One question. Is it possible to include the name of the nearest previous marker in the output filename? Using $MARKER will only use marker name on that specific frame. What I want is to use the marker name for all following frames until the next marker.

**Comment by p2or** [comment]
Hi @nenad-jalsovec,

at first glance, the following expression should do (make sure there is a marker on the first frame):

- `next(reversed([i.name for i in bpy.context.scene.timeline_markers if i.frame <= bpy.context.scene.frame_current]), "NO_MARKER")` 

Otherwise please open up a [new issue](https://github.com/p2or/blender-loom/issues/new) (as already [mentioned above](https://github.com/p2or/blender-loom/issues/11#issuecomment-1341282793)), thanks.

Cheers,
Christian

**Comment by nenad-jalsovec** [comment]
@p2or 

Yep, this works great! Thanks Christian.

I added a new global variable called $LAST_MARKER that runs the expression. Such a global should perhaps be included in the defaults. I think it's quite useful.

**Comment by p2or** [comment]
Hi @Fiztban,

as of [version 0.9.2](https://github.com/p2or/blender-loom/tree/0.9.2), _Loom_ adds basic support for the _Sequencer_ (I guess you already figured out that rendering scene strips was possible before by simply adding a 'fake camera' and rendering everything via command line). Given that opportunity to support the sequencer, I thought I could finally take care of your request. 

The bad news is that it would cause a lot of overhead while rendering, as we would have to e.g. 

- Recognize the active strip for all frames within the timeline of the sequencer
- Calculate the offset to each frame in the nested scene 
- Automatically replace the eferences in all given 'globals' (e.g. `bpy.context.scene` to `bpy.data.scene['SCENE']`) 
- And so on...

Means that it's pretty hard to support any nested (secondary) 'globals' within each scene strip. Good news is that, the warning is gone, the regular render won't be blocked anymore and you can use e.g. the markers in the main scene to format your filepath: 

![ezgif-7-9eb4750c79](https://github.com/p2or/blender-loom/assets/512368/c8b4ccd1-77b2-4c3f-94fb-027bddabc04e)

However, since you also want to use the file output node to store the images in different locations, I think another good option is to use [Batch Rendering](https://github.com/p2or/blender-loom/tree/0.9.2?tab=readme-ov-file#batch-render-dialog-ctrlshiftaltf12). Also, if you'd like to render multiple scenes of one file, you could make the scene you want to render the "active" scene and hit the snapshot button, which stores the current blend with all things set into a new folder called `temp` (can be changed in the preferences) and automatically adds it to the queue:

<img width="762" alt="Screenshot 2024-03-12 at 09 40 59" src="https://github.com/p2or/blender-loom/assets/512368/8cb9e274-fbf8-4109-bcb1-d606db49c96a">
<br><br>

I apologize if this is way too late, but I just wanted to let you know. Closing for now, can be reopened at any time. Cheers!
